### PR TITLE
[WIP] Add `assetPrefix` support

### PR DESCRIPTION
**Pull request by stephenmathieson**
This patch introduces an `assetPrefix` string to the global Next.js config. This will allow using a CDN to serve assets rather than Next.js itself.

A contrived example of using `assetPrefix` has been added as well.

Closes https://github.com/zeit/next.js/issues/737.

### CODE REVIEW

**Comment by arunoda** [comment]
I think we need to pass this via the next config. Because we need to use the prefix in the the lib/router/router as well when fetching JSON pages.

You could attach the config's prefix into NEXT_DATA and get it from the router.

**Comment by stephenmathieson** [comment]
Ah, yep, you're right! I'll give that a shot in a bit =)

File: examples/cdn-assets/next.config.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  assetPrefix: 'http://localhost:9999/'
```
Comment (arunoda) [inline examples/cdn-assets/next.config.js]: Add a comment saying something similar:

"In a real app, you need to put the CDN's URL prefix here"

**Comment by arunoda** [comment]
Overall, this looks great to me. I'll do a proper review today.

File: lib/router/router.js [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+import { parse, format, resolve as resolveURL } from 'url'
 import evalScript from '../eval-script'
 import shallowEquals from '../shallow-equals'
 import { EventEmitter } from 'events'
 import { reloadIfPrefetched } from '../prefetch'
 import { loadGetInitialProps } from '../utils'
 
 export default class Router extends EventEmitter {
-  constructor (pathname, query, { Component, ErrorComponent, err } = {}) {
+  constructor (pathname, query, { Component, ErrorComponent, err, assetPrefix } = {}) {
```
Comment (stephenmathieson) [inline lib/router/router.js]: Looks like we're using `global.__NEXT_DATA__` in this file, so I don't think I need to pass `assetPrefix` into the `Router` class.

Thoughts @arunoda?

File: server/config.js [hunk 3]
```diff
@@ -5,7 +5,8 @@ const cache = new Map()
 
 const defaultConfig = {
   webpack: null,
-  poweredByHeader: true
+  poweredByHeader: true,
+  assetPrefix: '/'
```
Comment (stephenmathieson) [inline server/config.js]: I'm wondering if this should be a function that returns the prefix. Something like:

```js
assetPrefix ({ buildId }) {
  return `/_next/${buildId}/`
}
```

File: lib/router/router.js (hunk 2 above)
Comment (arunoda) [inline lib/router/router.js]: Passing `assetPrefix` is a really good thing. We need to remove the direct `__NEXT_DATA__` use from this page. We could do it with a new PR.

Comment (stephenmathieson) [inline lib/router/router.js]: Do we want it in the "opts" object or as a direct param?

```js
constructor (pathname, query, { Component, ErrorComponent, err, assetPrefix } = {}) {
```

vs

```
constructor (pathname, query, assetPrefix, { Component, ErrorComponent, err } = {}) {
```

I'm not sure which is cleaner, but happy to refactor if you like.

File: server/config.js (hunk 3 above)
Comment (arunoda) [inline server/config.js]: I like this. I a real CDN setup, it's possible user is not using our `buildId`. And we could simply use something like `assetPrefix` and avoid the use of buildId directly in the client side.
(Becuase buildId is in the `assetPrefix`)

Yep. Implement this.

File: lib/router/router.js (hunk 2 above)
Comment (arunoda) [inline lib/router/router.js]: I'll also don't have strong feeling about this. So keep it like this. No need to change.

File: server/render.js [hunk 4]
```diff
@@ -66,7 +67,7 @@ async function doRender (req, res, pathname, query, {
       Component,
       props,
       err,
-      router: new Router(pathname, query)
+      router: new Router(pathname, query, { assetPrefix })
```
Comment (arunoda) [inline server/render.js]: `assetPrefix` is something we use only in the client. So, we don't need to pass it here.

File: lib/router/router.js (hunk 2 above)
Comment (stephenmathieson) [inline lib/router/router.js]: Cool, OK.

FWIW, the only usage of `global.__NEXT_DATA__` in this file is for the `buildId` property. We'll be able to remove the usage of the global in this PR =)

**Comment by stephenmathieson** [comment]
Looks like this needs a bit more work. The CDN needs to support the logic built-in [here](https://github.com/zeit/next.js/blob/2.0.0-beta.20/server/index.js#L88-L93), or we'll need to refactor the URL building stuff a bit more.

I'm -1 on making the CDN do anything other than serve assets, so I'll try to find some time to refactor.

Any advice would be hugely helpful =)

**Comment by arunoda** [comment]
In this case we need to fetch the `.json` using the exact URL. For an example instead of `/` we need to fetch `index.json` And for for `/abc` we need to fetch `/abc.json` or `/abc/index.json`. So, that's a problem.

I'll also look into that and see how to offload these stuff into the client. One option is to send a map of JSON files and their paths to the client with `NEXT_DATA`.

File: examples/cdn-assets/cdn.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const express = require('express')
+const path = require('path')
+
+const NEXT_DIR = path.join(__dirname, '.next')
+
+const app = express()
+// `-` is the default build ID. A real CDN would use a version number/etc here.
+app.use('/assets/-/', express.static(NEXT_DIR))
```
Comment (nkzawa) [inline examples/cdn-assets/cdn.js]: Is it difficult to make this example work on production too ?

File: lib/prefetch.js [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 import React from 'react'
 import Link, { isLocal } from './link'
-import { parse as urlParse } from 'url'
+import { parse as urlParse, resovle as urlResolve } from 'url'
```
Comment (satazor) [inline lib/prefetch.js]: resolve typo

**Comment by satazor** [comment]
This is great step forward! @stephenmathieson thanks for your work.

Regarding the PR: what's currently blocking this?

**Comment by timneutkens** [comment]
@satazor there's a few review points.

@stephenmathieson could you fix them? ❤️

**Comment by stephenmathieson** [comment]
@satazor 

> what's currently blocking this?

A few things. I'm pretty strapped for time right now unfortunately. I may have a few hours this weekend, but no promises =(

Also, this needs to be addressed: https://github.com/zeit/next.js/pull/907#issuecomment-275890064.

@timneutkens 

> could you fix them? 

Yes, of course! I've just been crazy busy lately. My apologies.

File: examples/cdn-assets/cdn.js (hunk 5 above)
Comment (stephenmathieson) [inline examples/cdn-assets/cdn.js]: Not hard, necessarily, but I don't think it's worth the effort. In production, I'll be using CloudFront, not a contrived node script.

Thoughts?

**Comment by timneutkens** [comment]
@stephenmathieson awesome! no rush on this ❤️

**Comment by arunoda** [comment]
Even though this PR seems like a simple change. But it involves a lot of changes specially with how we load chunks. (code splitted)

So, we'd like to defer this until we ship 2.0

**Comment by timneutkens** [comment]
@arunoda milestoned it 2.1

**Comment by rauchg** [comment]
I want to land this ASAP

**Comment by arunoda** [comment]
@rauchg this is the issue we are facing: https://github.com/zeit/next.js/pull/907#issuecomment-275890064

**Comment by arunoda** [comment]
@stephenmathieson Thanks for this PR.
I submit a pretty simple solution based on this PR.
Kudos to @rauchg 

See: #1700

**Pull request closed without merging**